pepper-apis/src/main/java/org/broadinstitute/ddp/route/PutFormAnswersRoute.java

```diff
@@ -226,6 +229,48 @@ public Object handle(Request request, Response response) {
         return resp;
     }
 
+    /**
+     * If this is the first submit for the activity instance and should snapshot flag is true,
+     * then save a snapshot of substitutions.
+     */
+    private void snapshotSubstitutions(Handle handle, String studyGuid, String operatorGuid,
+                                       FormInstance form, boolean shouldSnapshotSubstitutionsOnSubmit) {
+        if (shouldSnapshotSubstitutionsOnSubmit) {
+            handle.attach(ActivityInstanceDao.class).saveSubstitutions(
+                    form.getInstanceId(),
+                    I18nContentRenderer.newValueProvider(
+                            handle, form.getParticipantUserId(), operatorGuid, studyGuid).getSnapshot());
+        }
+    }
+
+    /**
+     * Snapshot address on submit.
+     * If flag shouldSnapshotAddressOnSubmit is true then snapshot an address.
+     *
+     * <p>Algorithm:
+     * <ul>
+     *     <li>Find the user’s “default” address;</li>
+     *     <li>Create a copy of the address but don’t mark it as the default;</li>
+     *     <li>Save the address guid as “snapshot” variable I18nTemplateConstants.Snapshot.ADDRESS_GUID.</li>
+     * </ul>
+     */
+    private void snapshotAddress(Handle handle, String participantGuid, String operatorGuid, FormInstance form,
+                                 boolean shouldSnapshotAddressOnSubmit) {
+        if (shouldSnapshotAddressOnSubmit) {
+            // find default address
+            Optional<MailAddress> defaultAddress = addressService.findDefaultAddressForParticipant(handle, participantGuid);
+            if (defaultAddress.isPresent()) {
+                MailAddress mailAddress = defaultAddress.get();
+                mailAddress.setDefault(false);
+                // create a copy of the default address (but setting it as non-default)
+                mailAddress = addressService.addAddress(handle, mailAddress, participantGuid, operatorGuid);
```

Comment:
If you look at the implementation of `AddressService.addAddress()`, under the hood it's doing a lot of things: calculating plus codes, calling EasyPost API to verify address, etc. Those are nice things, but we don't need them in this case. The default address should already be in a valid state, so we should just "copy" it instead of doing all this extra work. Is there something else we can use here?

Reply:
Done